src/evagram/website/backend/api/tests.py

```diff
@@ -0,0 +1,160 @@
+
+    def test_initial_load(self):
+        response = self.client.get("/api/initial-load/")
+        self.assertEqual(200, response.status_code)
+        self.assertTrue("owners" in response.json())
+        self.assertTrue("experiments" in response.json())
+        self.assertTrue("groups" in response.json())
+        self.assertTrue("observations" in response.json())
+        self.assertTrue("variables" in response.json())
+
+    def test_get_plot_components(self):
+        response = self.client.get(
+            "/api/get-plot-components/?experiment_id=12&observation_id=1&variable_id=1&group_id=1")
+        self.assertEqual("experiment_id=12&observation_id=1&variable_id=1&group_id=1",
+                         response.request['QUERY_STRING'])
+        self.assertEqual(200, response.status_code)
+        # check if plot components in response match with plot id in database
+        plot = Plots.objects.get(pk=response.json()["plot_id"])
+        self.assertEqual(plot.div, response.json()["div"])
+        self.assertEqual(plot.script, response.json()["script"])
+
+    def test_plot_insufficient_params(self):
+        response = self.client.get("/api/get-plot-components/?experiment_id=12&observation_id=1")
+        self.assertEqual("experiment_id=12&observation_id=1", response.request['QUERY_STRING'])
+        self.assertEqual(400, response.status_code)
+        self.assertEqual("Missing request parameter detected: 'variable_id'",
+                         response.json()['error'])
+
+    def test_plot_invalid_plot_fks(self):
+        response = self.client.get(
+            "/api/get-plot-components/?experiment_id=-12&observation_id=1&variable_id=1&group_id=1")
+        self.assertEqual("experiment_id=-12&observation_id=1&variable_id=1&group_id=1",
+                         response.request['QUERY_STRING'])
+        self.assertEqual(400, response.status_code)
+        self.assertEqual("Plots matching query does not exist.", response.json()['error'])
+
+    def test_plot_bad_input(self):
+        response = self.client.get(
+            "/api/get-plot-components/?experiment_id=xyz&observation_id=1&variable_id=1&group_id=1")
+        self.assertEqual("experiment_id=xyz&observation_id=1&variable_id=1&group_id=1",
+                         response.request['QUERY_STRING'])
+        self.assertEqual(400, response.status_code)
+        self.assertEqual("Field 'experiment_id' expected a number but got 'xyz'.",
+                         response.json()['error'])
+
+    def test_update_user_valid_pk(self):
+        response = self.client.get("/api/update-user-option/?owner_id=1")
+        self.assertEqual("owner_id=1", response.request['QUERY_STRING'])
+        self.assertEqual(200, response.status_code)
+        self.assertTrue("experiments" in response.json())
+        self.assertTrue("observations" in response.json())
+        self.assertTrue("variables" in response.json())
+        self.assertTrue("groups" in response.json())
```

Comment:
This is all fine, but I wonder if there's a cleaner way to write the `_update_` tests.  There is quite a bit of redundancy between user, experiment, observation, variable tests.

Reply:
The `__update__` tests are pretty redundant as they are testing the same input cases across user, experiment, observation and variables. My idea was to improve code coverage for all the API endpoints even though much of the `update_` views have the same control flow.